### DISCUSSION TITLE
match until/til?

### DISCUSSION THREAD

**Discussion by hh9527** [osa1/lexgen#58] [Q&A]
for example, how to lex rust like raw string: r###"..."###

```rust

rule Init {
  'r' '#'* '"' => |lexer| {
    *lexer.state() = lexer.match_().len() - 1; // set the number of end mark “#...
    lexer.switch_and_return(LexerRule::RawText, Token::RawTextStart)
  },
}

rule RawText {
  // here, I dont know how to constraint to number of '#'
  _* '"' '#'* => |lexer| {
    len n = *lexer.state();
    let cont = lexer.match_();
    let body = &cont[..cont.len() - n];
    lexer.switch_and_return(LexerRule::Init, Token::RawTextCont(body, n));
  },
}

```

**Comment by osa1** [comment]
Counting needs to be done using states. Here are two examples:

- Lexing Rust `r###...` syntax: https://github.com/osa1/lexgen_rust/blob/9e42b50572a2be12509d458e9be4d663c5fc4fe8/crates/lexgen_rust/src/lib.rs#L475-L479
- Lexing Lua `[==[...` syntax: https://github.com/osa1/lexgen/blob/271ba43edecacc2d20155d293a1354b8fd9159d3/crates/lexgen/tests/lua_5_1.rs#L189-L198 a simpler version of this is in the README: https://github.com/osa1/lexgen/#stateful-lexer-example
